server/src/main/java/com/project/yozmcafe/service/CafeHistoryService.java

```diff
@@ -0,0 +1,47 @@
+package com.project.yozmcafe.service;
+
+import java.util.Collections;
+import java.util.List;
+
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+import com.project.yozmcafe.domain.cafe.Cafe;
+import com.project.yozmcafe.domain.cafe.CafeRepository;
+import com.project.yozmcafe.domain.cafe.UnViewedCafe;
+import com.project.yozmcafe.domain.cafe.UnViewedCafeRepository;
+import com.project.yozmcafe.domain.member.Member;
+
+@Service
+@Transactional(readOnly = true)
+public class CafeHistoryService {
+
+    private final CafeRepository cafeRepository;
+    private final UnViewedCafeRepository unViewedCafeRepository;
+
+    public CafeHistoryService(final CafeRepository cafeRepository, final UnViewedCafeRepository unViewedCafeRepository) {
+        this.cafeRepository = cafeRepository;
+        this.unViewedCafeRepository = unViewedCafeRepository;
+    }
+
+    @Transactional
+    public void removeUnViewedCafe(final Member member, final long cafeId) {
+        final Cafe cafe = cafeRepository.findById(cafeId)
+                .orElseThrow(() -> new IllegalArgumentException("존재하지 않는 카페입니다."));
+        member.removeUnViewedCafe(cafe);
+
+        if (member.isEmptyUnViewedCafe()) {
+            updateUnViewedCafes(member);
+        }
+    }
+
+    private void updateUnViewedCafes(final Member member) {
+        final List<Cafe> allCafes = cafeRepository.findAll();
+        Collections.shuffle(allCafes);
+        final List<UnViewedCafe> allUnViewedCafes = allCafes.stream()
+                .map(savedCafe -> new UnViewedCafe(null, savedCafe, member))
+                .toList();
+        this.unViewedCafeRepository.saveAll(allUnViewedCafes);
+        member.addUnViewedCafes(allUnViewedCafes);
+    }
```

Comment:
```suggestion
        final List<Cafe> allCafes = cafeRepository.findAll();
        Collections.shuffle(allCafes);
        member.updateUnViewedCafes(allCafes);
```
1. CASCADE가 ALL이기 때문에 위와 같은 형태로 바꿔도 되지 않을까요?
2. UnViewedCafe 생성자에서 ID 부분에 null을 직접 넣는 것보다 생성자 오버로딩이 나을 것 같아요

Reply:
1. 해당 메서드에서 `this.unViewedCafeRepository.saveAll(allUnViewedCafes);` 이부분은 cascade.all 로 인해 불필요한 호출이 될것 같아 삭제했습니다.
Member 객체에서 List<UnViewedCafe> 를 필드로 갖고 있어 Cafe -> UnViewedCafe로 매핑하는 부분은 남겨뒀어요~

3. 반영했습니다!

Reply:
casacade = all 일때는 member에 unviewCafe가 들어가면 영속성이 이어져 db에 저장되겠네요. 맞는 말입니다. 반영하겠습니다. 생성자도 null은 오버로딩한 생성자를 통해 넣어주는 게 낫겠네요.